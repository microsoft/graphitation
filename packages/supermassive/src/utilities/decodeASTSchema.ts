import {
  DirectiveDefinitionNode,
  DocumentNode,
  EnumTypeDefinitionNode,
  FieldDefinitionNode,
  InputObjectTypeDefinitionNode,
  InputValueDefinitionNode,
  InterfaceTypeDefinitionNode,
  Kind,
  ListTypeNode,
  NamedTypeNode,
  NameNode,
  NonNullTypeNode,
  ObjectTypeDefinitionNode,
  ScalarTypeDefinitionNode,
  UnionTypeDefinitionNode,
} from "graphql";
import {
  DirectiveDefinitionTuple,
  SchemaDefinitions,
  EnumTypeDefinitionTuple,
  FieldDefinition,
  InputObjectTypeDefinitionTuple,
  InputValueDefinition,
  InterfaceTypeDefinitionTuple,
  ObjectTypeDefinitionTuple,
  TypeDefinitionsRecord,
  UnionTypeDefinitionTuple,
  isScalarTypeDefinition,
  isEnumTypeDefinition,
  isObjectTypeDefinition,
  isInterfaceTypeDefinition,
  isUnionTypeDefinition,
  isInputObjectTypeDefinition,
  getEnumValues,
  getObjectFields,
  getObjectTypeInterfaces,
  getFields,
  getInterfaceTypeInterfaces,
  getUnionTypeMembers,
  getInputObjectFields,
  getFieldTypeReference,
  getFieldArgs,
  getInputValueTypeReference,
  getInputDefaultValue,
  getDirectiveName,
  getDirectiveDefinitionArgs,
  getDirectiveLocations,
  decodeDirectiveLocation,
} from "../schema/definition";
import {
  inspectTypeReference,
  isListType,
  isNonNullType,
  typeNameFromReference,
  TypeReference,
  unwrap,
} from "../schema/reference";
import { invariant } from "../jsutils/invariant";
import { ValueNode as ConstValueNode } from "graphql/language/ast"; // TODO: use ConstValueNode in graphql@17
import { inspect } from "../jsutils/inspect";

/**
 * Converts encoded schema to standard AST representation of the same schema
 */
export function decodeASTSchema(
  encodedSchemaFragments: SchemaDefinitions[],
): DocumentNode {
  if (encodedSchemaFragments.length !== 1) {
    // TODO:
    throw new Error("decodeSchema does not support decoding extensions yet");
  }
  const definitions = [];
  const types = encodedSchemaFragments[0].types;
  const directives = encodedSchemaFragments[0].directives;

  for (const typeName in types) {
    const tuple = types[typeName];
    if (isScalarTypeDefinition(tuple)) {
      definitions.push(decodeScalarType(typeName));
    } else if (isEnumTypeDefinition(tuple)) {
      definitions.push(decodeEnumType(typeName, tuple));
    } else if (isObjectTypeDefinition(tuple)) {
      definitions.push(decodeObjectType(typeName, tuple, types));
    } else if (isInterfaceTypeDefinition(tuple)) {
      definitions.push(decodeInterfaceType(typeName, tuple, types));
    } else if (isUnionTypeDefinition(tuple)) {
      definitions.push(decodeUnionType(typeName, tuple));
    } else if (isInputObjectTypeDefinition(tuple)) {
      definitions.push(decodeInputObjectType(typeName, tuple, types));
    }
  }

  for (const directive of directives ?? []) {
    definitions.push(decodeDirective(directive, types));
  }

  return { kind: Kind.DOCUMENT, definitions };
}

function nameNode(value: string): NameNode {
  return { kind: Kind.NAME, value };
}

function decodeScalarType(typeName: string): ScalarTypeDefinitionNode {
  return {
    kind: Kind.SCALAR_TYPE_DEFINITION,
    name: nameNode(typeName),
  };
}

function decodeEnumType(
  typeName: string,
  tuple: EnumTypeDefinitionTuple,
): EnumTypeDefinitionNode {
  return {
    kind: Kind.ENUM_TYPE_DEFINITION,
    name: nameNode(typeName),
    values: getEnumValues(tuple).map((value) => ({
      kind: Kind.ENUM_VALUE_DEFINITION,
      name: nameNode(value),
    })),
  };
}

function decodeObjectType(
  typeName: string,
  tuple: ObjectTypeDefinitionTuple,
  types: TypeDefinitionsRecord,
): ObjectTypeDefinitionNode {
  return {
    kind: Kind.OBJECT_TYPE_DEFINITION,
    name: nameNode(typeName),
    fields: decodeFields(getObjectFields(tuple) ?? {}, types),
    interfaces: getObjectTypeInterfaces(tuple).map((name) => ({
      kind: Kind.NAMED_TYPE,
      name: nameNode(name),
    })),
  };
}

function decodeInterfaceType(
  typeName: string,
  tuple: InterfaceTypeDefinitionTuple,
  types: TypeDefinitionsRecord,
): InterfaceTypeDefinitionNode {
  return {
    kind: Kind.INTERFACE_TYPE_DEFINITION,
    name: nameNode(typeName),
    fields: decodeFields(getFields(tuple), types),
    interfaces: getInterfaceTypeInterfaces(tuple).map((name) => ({
      kind: Kind.NAMED_TYPE,
      name: nameNode(name),
    })),
  };
}

function decodeUnionType(
  typeName: string,
  tuple: UnionTypeDefinitionTuple,
): UnionTypeDefinitionNode {
  return {
    kind: Kind.UNION_TYPE_DEFINITION,
    name: nameNode(typeName),
    types: getUnionTypeMembers(tuple).map((name) => ({
      kind: Kind.NAMED_TYPE,
      name: nameNode(name),
    })),
  };
}

function decodeInputObjectType(
  typeName: string,
  tuple: InputObjectTypeDefinitionTuple,
  types: TypeDefinitionsRecord,
): InputObjectTypeDefinitionNode {
  return {
    kind: Kind.INPUT_OBJECT_TYPE_DEFINITION,
    name: nameNode(typeName),
    fields: Object.entries(getInputObjectFields(tuple)).map(([name, value]) =>
      decodeInputValue(name, value, types),
    ),
  };
}

function decodeFields(
  fields: Record<string, FieldDefinition>,
  types: TypeDefinitionsRecord,
): FieldDefinitionNode[] {
  return Object.entries(fields).map(([name, value]) => {
    const type = decodeTypeReference(getFieldTypeReference(value));
    return {
      kind: Kind.FIELD_DEFINITION,
      name: nameNode(name),
      type,
      arguments: decodeArguments(getFieldArgs(value) ?? {}, types),
    };
  });
}

function decodeInputValue(
  name: string,
  value: InputValueDefinition,
  types: TypeDefinitionsRecord,
): InputValueDefinitionNode {
  const inputValueTypeRef = getInputValueTypeReference(value);
  const type = decodeTypeReference(inputValueTypeRef);
  return {
    kind: Kind.INPUT_VALUE_DEFINITION,
    name: nameNode(name),
    type,
    defaultValue: Array.isArray(value)
      ? (valueToConstValueNode(
          getInputDefaultValue(value),
          inputValueTypeRef,
          types,
        ) as any) // Note: "any" is necessary here for graphql15/graphql17 cross-compatibility
      : undefined,
  };
}

function valueToConstValueNode(
  jsValue: unknown,
  typeRef: TypeReference,
  types: TypeDefinitionsRecord,
): ConstValueNode {
  const typeName = typeNameFromReference(typeRef);
  if (typeof jsValue === "string") {
    return typeName === "String"
      ? { kind: Kind.STRING, value: jsValue }
      : { kind: Kind.ENUM, value: jsValue };
  }
  if (jsValue === null) {
    invariant(!isNonNullType(typeRef), "Expecting nullable type");
    return { kind: Kind.NULL };
  }
  if (Number.isInteger(jsValue)) {
    return { kind: Kind.INT, value: String(jsValue) };
  }
  if (typeof jsValue === "boolean") {
    return { kind: Kind.BOOLEAN, value: jsValue };
  }
  if (typeof jsValue === "number") {
    return { kind: Kind.FLOAT, value: String(jsValue) };
  }
  if (Array.isArray(jsValue)) {
    invariant(isListType(typeRef), "Expecting list type");
    return {
      kind: Kind.LIST,
      values: jsValue.map((item) =>
        valueToConstValueNode(item, unwrap(typeRef), types),
      ),
    };
  }
  if (typeof jsValue === "object") {
    const typeDef = types[typeName];
    invariant(
      Array.isArray(typeDef) && isInputObjectTypeDefinition(typeDef),
      `Expecting input object type for ${typeName}, got ${typeDef?.[0]}`,
    );
    const fields = getInputObjectFields(typeDef);
    return {
      kind: Kind.OBJECT,
      fields: Object.entries(jsValue).map(([name, value]) => {
        const fieldDef = fields[name];
        const fieldTypeRef = getInputValueTypeReference(fieldDef);
        invariant(
          fieldTypeRef !== undefined,
          `Could not find field definition for ${typeName}.${name}`,
        );
        return {
          kind: Kind.OBJECT_FIELD,
          name: nameNode(name),
          value: valueToConstValueNode(value, fieldTypeRef, types),
        };
      }),
    };
  }
  invariant(
    false,
    `Unexpected value for type ${inspectTypeReference(typeRef)}: ${inspect(
      jsValue,
    )}`,
  );
}

function decodeArguments(
  args: Record<string, InputValueDefinition>,
  types: TypeDefinitionsRecord,
): InputValueDefinitionNode[] {
  return Object.entries(args).map(([name, value]) =>
    decodeInputValue(name, value, types),
  );
}

function decodeTypeReference(
  ref: TypeReference,
): NamedTypeNode | ListTypeNode | NonNullTypeNode {
  if (isListType(ref)) {
    return {
      kind: Kind.LIST_TYPE,
      type: decodeTypeReference(unwrap(ref)),
    };
  }
  if (isNonNullType(ref)) {
    const unwrappedType = decodeTypeReference(unwrap(ref));
    invariant(
      unwrappedType.kind !== Kind.NON_NULL_TYPE,
      "Non-null modifier cannot be nested",
    );
    return {
      kind: Kind.NON_NULL_TYPE,
      type: unwrappedType,
    };
  }
  return {
    kind: Kind.NAMED_TYPE,
    name: nameNode(typeNameFromReference(ref)),
  };
}

function decodeDirective(
  tuple: DirectiveDefinitionTuple,
  types: TypeDefinitionsRecord,
): DirectiveDefinitionNode {
  const name = getDirectiveName(tuple);
  const args = getDirectiveDefinitionArgs(tuple);
  const locations = getDirectiveLocations(tuple);
  return {
    kind: Kind.DIRECTIVE_DEFINITION,
    name: nameNode(name),
    arguments: args ? decodeArguments(args, types) : [],
    locations: locations.map((loc) => ({
      kind: Kind.NAME,
      value: decodeDirectiveLocation(loc),
    })),
    // TODO? repeatable are irrelevant for execution
    repeatable: false,
  };
}
