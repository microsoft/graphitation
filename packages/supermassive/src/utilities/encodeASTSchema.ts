import {
  DirectiveDefinitionNode,
  EnumTypeDefinitionNode,
  FieldDefinitionNode,
  InputObjectTypeDefinitionNode,
  InputValueDefinitionNode,
  InterfaceTypeDefinitionNode,
  ObjectTypeDefinitionNode,
  ScalarTypeDefinitionNode,
  UnionTypeDefinitionNode,
  DocumentNode,
} from "graphql";
import {
  DirectiveDefinitionTuple,
  EnumTypeDefinitionTuple,
  FieldDefinitionTuple,
  InputObjectTypeDefinitionTuple,
  InputValueDefinitionRecord,
  InputValueDefinitionTuple,
  InterfaceTypeDefinitionTuple,
  ObjectTypeDefinitionTuple,
  ScalarTypeDefinitionTuple,
  UnionTypeDefinitionTuple,
  SchemaFragmentDefinitions,
  TypeKind,
  TypeReference,
} from "../schema/definition";
import { typeReferenceFromNode } from "../schema/reference";
import { valueFromASTUntyped } from "./valueFromASTUntyped";

export function encodeASTSchema(
  schemaFragment: DocumentNode,
): SchemaFragmentDefinitions {
  const types = Object.create(null);
  const directives = [];
  for (const definition of schemaFragment.definitions) {
    if (definition.kind === "ObjectTypeDefinition") {
      types[definition.name.value] = encodeObjectType(definition);
    } else if (definition.kind === "InputObjectTypeDefinition") {
      types[definition.name.value] = encodeInputObjectType(definition);
    } else if (definition.kind === "EnumTypeDefinition") {
      types[definition.name.value] = encodeEnumType(definition);
    } else if (definition.kind === "UnionTypeDefinition") {
      types[definition.name.value] = encodeUnionType(definition);
    } else if (definition.kind === "InterfaceTypeDefinition") {
      types[definition.name.value] = encodeInterfaceType(definition);
    } else if (definition.kind === "ScalarTypeDefinition") {
      types[definition.name.value] = encodeScalarType(definition);
    } else if (definition.kind === "DirectiveDefinition") {
      directives.push(encodeDirective(definition));
    }
  }
  return !directives.length ? { types } : { types, directives };
}

function encodeScalarType(
  _type: ScalarTypeDefinitionNode,
): ScalarTypeDefinitionTuple {
  return [TypeKind.SCALAR];
}

function encodeEnumType(node: EnumTypeDefinitionNode): EnumTypeDefinitionTuple {
  return [TypeKind.ENUM, (node.values ?? []).map((value) => value.name.value)];
}

function encodeObjectType(
  node: ObjectTypeDefinitionNode,
): ObjectTypeDefinitionTuple {
  const fields = Object.create(null);
  for (const field of node.fields ?? []) {
    fields[field.name.value] = encodeField(field);
  }
  if (!node.interfaces) {
    return [TypeKind.OBJECT, fields];
  }
  return [
    TypeKind.OBJECT,
    fields,
    node.interfaces.map((iface) => iface.name.value),
  ];
}

function encodeInterfaceType(
  node: InterfaceTypeDefinitionNode,
): InterfaceTypeDefinitionTuple {
  const fields = Object.create(null);
  for (const field of node.fields ?? []) {
    fields[field.name.value] = encodeField(field);
  }
  if (!node.interfaces) {
    return [TypeKind.INTERFACE, fields];
  }
  return [
    TypeKind.INTERFACE,
    fields,
    node.interfaces.map((iface) => iface.name.value),
  ];
}

function encodeUnionType(
  node: UnionTypeDefinitionNode,
): UnionTypeDefinitionTuple {
  return [TypeKind.UNION, (node.types ?? []).map((type) => type.name.value)];
}

function encodeInputObjectType(
  node: InputObjectTypeDefinitionNode,
): InputObjectTypeDefinitionTuple {
  const fields = Object.create(null);
  for (const field of node.fields ?? []) {
    fields[field.name.value] = encodeInputValue(field);
  }
  return [TypeKind.INPUT, fields];
}

function encodeField(
  node: FieldDefinitionNode,
): TypeReference | FieldDefinitionTuple {
  return !node.arguments?.length
    ? typeReferenceFromNode(node.type)
    : [typeReferenceFromNode(node.type), encodeArguments(node)];
}

function encodeArguments(
  node: FieldDefinitionNode | DirectiveDefinitionNode,
): InputValueDefinitionRecord {
  const args = Object.create(null);
  for (const argument of node.arguments ?? []) {
    args[argument.name.value] = encodeInputValue(argument);
  }
  return args;
}

function encodeInputValue(
  node: InputValueDefinitionNode,
): InputValueDefinitionTuple | TypeReference {
  if (!node.defaultValue) {
    return typeReferenceFromNode(node.type);
  }
  return [
    typeReferenceFromNode(node.type),
    valueFromASTUntyped(node.defaultValue),
  ];
}

function encodeDirective(
  node: DirectiveDefinitionNode,
): DirectiveDefinitionTuple {
  return !node.arguments?.length
    ? [node.name.value]
    : [node.name.value, encodeArguments(node)];
}
