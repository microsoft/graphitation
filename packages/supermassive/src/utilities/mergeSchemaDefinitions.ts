import {
  DirectiveDefinitionTuple,
  FieldDefinitionRecord,
  getDirectiveDefinitionArgs,
  getDirectiveName,
  getFieldArgs,
  getFields,
  getInputObjectFields,
  InputValueDefinitionRecord,
  isInputObjectTypeDefinition,
  isInterfaceTypeDefinition,
  isObjectTypeDefinition,
  SchemaDefinitions,
  setDirectiveDefinitionArgs,
  setFieldArgs,
  TypeDefinitionsRecord,
} from "../schema/definition";
import { inspect } from "../jsutils/inspect";

export function mergeSchemaDefinitions(
  accumulator: SchemaDefinitions,
  definitions: SchemaDefinitions[],
): SchemaDefinitions {
  if (!definitions.length) {
    return accumulator;
  }
  for (const source of definitions) {
    if (!accumulator.types) {
      accumulator.types = source.types;
    } else if (source.types) {
      mergeTypes(accumulator.types, source.types);
    }
    if (!accumulator.directives) {
      accumulator.directives = source.directives;
    } else if (source.directives?.length) {
      mergeDirectives(accumulator.directives, source.directives);
    }
  }
  return accumulator;
}

export function mergeDirectives(
  target: DirectiveDefinitionTuple[],
  source: DirectiveDefinitionTuple[],
) {
  for (const sourceDirective of source) {
    const targetDirective = target.find(
      (directive) =>
        getDirectiveName(directive) === getDirectiveName(sourceDirective),
    );
    if (!targetDirective) {
      target.push(sourceDirective);
      continue;
    }
    const sourceArgs = getDirectiveDefinitionArgs(sourceDirective);
    if (!sourceArgs) {
      continue;
    }
    const targetArgs = getDirectiveDefinitionArgs(targetDirective);
    if (!targetArgs) {
      setDirectiveDefinitionArgs(targetDirective, sourceArgs);
      continue;
    }
    mergeInputValues(targetArgs, sourceArgs);
  }
}

/**
 * Adds missing definitions from source into target. Mutates target in place.
 */
export function mergeTypes(
  target: TypeDefinitionsRecord,
  source: TypeDefinitionsRecord,
): void {
  for (const [typeName, sourceDef] of Object.entries(source)) {
    const targetDef = target[typeName];
    if (!targetDef) {
      target[typeName] = sourceDef;
      continue;
    }
    if (
      (isObjectTypeDefinition(targetDef) &&
        isObjectTypeDefinition(sourceDef)) ||
      (isInterfaceTypeDefinition(targetDef) &&
        isInterfaceTypeDefinition(sourceDef))
    ) {
      mergeFields(getFields(targetDef), getFields(sourceDef));
      // Note: not merging implemented interfaces - assuming they are fully defined by the first occurrence
      continue;
    }
    if (
      isInputObjectTypeDefinition(targetDef) &&
      isInputObjectTypeDefinition(sourceDef)
    ) {
      mergeInputValues(
        getInputObjectFields(targetDef),
        getInputObjectFields(sourceDef),
      );
      continue;
    }
    // Note: not merging scalars, unions and enums - assuming they are fully defined by the first occurrence
    if (targetDef[0] !== sourceDef[0]) {
      throw new Error(
        `Type ${typeName} is represented differently in different schema fragments:\n` +
          inspect(targetDef) +
          `\n` +
          inspect(sourceDef),
      );
    }
  }
}

function mergeFields(
  target: FieldDefinitionRecord,
  source: FieldDefinitionRecord,
) {
  for (const [fieldName, sourceDef] of Object.entries(source)) {
    const targetDef = target[fieldName];
    if (!target[fieldName] || !Array.isArray(targetDef)) {
      target[fieldName] = sourceDef;
      continue;
    }
    const sourceArgs = getFieldArgs(sourceDef);
    if (sourceArgs) {
      const targetArgs = getFieldArgs(targetDef) ?? setFieldArgs(targetDef, {});
      mergeInputValues(targetArgs, sourceArgs);
    }
  }
}

function mergeInputValues(
  target: InputValueDefinitionRecord,
  source: InputValueDefinitionRecord,
) {
  for (const [fieldName, sourceDef] of Object.entries(source)) {
    const targetDef = target[fieldName];
    if (!target[fieldName] || !Array.isArray(targetDef)) {
      target[fieldName] = sourceDef;
    }
    // Note: not merging defaultValue - assuming it is fully defined by the first occurrence
  }
}
