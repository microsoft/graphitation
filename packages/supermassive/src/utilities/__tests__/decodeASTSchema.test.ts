import { DocumentNode, visit, print } from "graphql";
import { decodeASTSchema } from "../decodeASTSchema";
import { encodeASTSchema } from "../encodeASTSchema";
import { kitchenSinkSDL } from "./fixtures/kitchenSinkSDL";
import { swapiSDL } from "./fixtures/swapiSDL";
import { mergeSchemaDefinitions } from "../mergeSchemaDefinitions";
import { schemaWithBooleanParameter } from "./fixtures/schemaWithBooleanParameter";

describe(decodeASTSchema, () => {
  test("correctly encodes swapi AST schema", () => {
    const doc = cleanUpDocument(swapiSDL.document);
    const encoded = encodeASTSchema(doc);
    const decoded = decodeASTSchema(encoded);

    expect(decoded).toEqual(doc);
    expect(encodeASTSchema(decoded)).toEqual(encoded);
    expect(print(decoded)).toMatchSnapshot();
  });

  test("correctly encodes kitchen sink AST schema", () => {
    const doc = cleanUpDocument(kitchenSinkSDL.document);
    const encoded = [
      mergeSchemaDefinitions(
        { types: {}, directives: [] },
        encodeASTSchema(doc),
      ),
    ];
    const decoded = decodeASTSchema(encoded);
    expect(encodeASTSchema(decoded)).toEqual(encoded);
    expect(print(decoded)).toMatchSnapshot();
  });

  test("correctly encodes a schema with a Boolean parameter", () => {
    const doc = cleanUpDocument(schemaWithBooleanParameter.document);
    const encoded = encodeASTSchema(doc);
    const decoded = decodeASTSchema(encoded);

    expect(decoded).toEqual(doc);
    expect(encodeASTSchema(decoded)).toEqual(encoded);
    expect(print(decoded)).toMatchSnapshot();
  });
});

function cleanUpDocument(doc: DocumentNode): DocumentNode {
  return visit(doc, {
    enter(node: any) {
      delete node.description;
      delete node.loc;
      delete node.block;
      if (node.directives?.length === 0) {
        delete node.directives;
      }
      return node;
    },
  });
}
