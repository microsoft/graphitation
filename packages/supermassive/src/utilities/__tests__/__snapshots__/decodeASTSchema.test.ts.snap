// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decodeASTSchema correctly encodes a schema with a Boolean parameter 1`] = `
"type Query {
  person(id: Int!): Person
}

type Person {
  name: String!
  phones(includeDefault: Boolean = true): [String!]!
}
"
`;

exports[`decodeASTSchema correctly encodes kitchen sink AST schema 1`] = `
"type Foo implements Bar & Baz & Two {
  one: Type
  two(argument: InputType!): Type
  three(argument: InputType, other: String): Int
  four(argument: String = "string"): String
  five(argument: [String] = ["string", "string"]): String
  six(argument: InputType = {key: "value"}): Type
  seven(argument: Int = null): Type
  eight(argument: OneOfInputType): Type
}

type AnnotatedObject {
  annotatedField(arg: Type = default): Type
}

type UndefinedType

interface Bar {
  one: Type
  four(argument: String = "string"): String
  two(argument: InputType!): Type
}

interface AnnotatedInterface {
  annotatedField(arg: Type): Type
}

interface UndefinedInterface

interface Baz implements Bar & Two {
  one: Type
  two(argument: InputType!): Type
  four(argument: String = "string"): String
}

union Feed = Story | Article | Advert

union AnnotatedUnion = A | B

union AnnotatedUnionTwo = A | B

union UndefinedUnion

scalar CustomScalar

scalar AnnotatedScalar

enum Site {
  DESKTOP
  MOBILE
  WEB
}

enum AnnotatedEnum {
  ANNOTATED_VALUE
  OTHER_VALUE
}

enum UndefinedEnum

input InputType {
  key: String!
  answer: Int = 42
  other: Float = 12300
}

input OneOfInputType {
  string: String
  int: Int
}

input AnnotatedInput {
  annotatedField: Type
}

input UndefinedInput

directive @skip(if: Boolean!) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT

directive @include(if: Boolean!) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT

directive @include2(if: Boolean!) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT

directive @myRepeatableDir(name: String!) on OBJECT | INTERFACE
"
`;

exports[`decodeASTSchema correctly encodes swapi AST schema 1`] = `
"union SearchResult = Person | Starship | Transport | Species | Vehicle | Planet | Film

enum NodeType {
  Person
  Starship
  Transport
  Species
  Vehicle
  Planet
  Film
}

type Subscription {
  emitPersons(limit: Int!, throwError: Boolean): Person
}

type Query {
  node(nodeType: NodeType!, id: Int!): Node
  search(search: String): [SearchResult]
  person(id: Int!): Person
  planet(id: Int!): Planet
  film(id: Int!): Film
  transport(id: Int!): Transport
  starship(id: Int!): Starship
  vehicle(id: Int!): Vehicle
  searchPeopleByName(search: String!): [Person]
  searchStarshipsByName(search: String!): [Starship]
  searchTransportsByName(search: String!): [Transport]
  searchSpeciesByName(search: String!): [Species]
  searchVehiclesByName(search: String!): [Vehicle]
  searchPlanetsByName(search: String!): [Planet]
  searchFilmsByTitle(search: String! = "The Empire Strikes Back"): [Film]
  allFilms: [Film]
  allStarships: [Starship]
  allPeople: [Person]
  allPlanets: [Planet]
  allSpecies: [Species]
  allTransports: [Transport]
  advancedDefaultInput(input: AdvancedInput! = {enumField: Transport, otherField: "Foo"}): String
  multiArger(a: Int, b: String, c: AdvancedInput): String
}

interface Node {
  id: Int
}

union Alive = Person | Species

type Film implements Node {
  title: String!
  starships: [Starship]
  edited: String
  vehicles: [Vehicle]
  planets: [Planet]
  producer: String
  created: String
  episode_id: Int
  director: String
  release_date: String
  opening_crawl: String
  characters: [Person]
  species: [Species]
  id: Int
}

type Vehicle implements Node {
  id: Int
  name: String
  vehicle_class: String
  pilots: [Person]
  edited: String
  consumables: String
  created: String
  model: String
  manufacturer: String
  image: String
  cargo_capacity: Int
  passengers: Int
  max_atmosphering_speed: Int
  crew: Int
  length: Float
  cost_in_credits: Int
}

type Person implements Node {
  id: Int
  edited: String
  name: String
  created: String
  gender: String
  skin_color: String
  hair_color: String
  height: Int
  eye_color: String
  mass: Int
  homeworld: Planet
  birth_year: String
  image: String
  vehicles: [Vehicle]
  starships: [Starship]
  films: [Film]
}

type Starship implements Node {
  id: Int
  films: [Film]
  pilots: [Person]
  MGLT: Int
  starship_class: String
  hyperdrive_rating: Float
  edited: String
  consumables: String
  name: String
  created: String
  cargo_capacity: Int
  passengers: Int
  max_atmosphering_speed: Int
  crew: String
  length: Int
  model: String
  cost_in_credits: Int
  manufacturer: String
  image: String
}

type Planet implements Node {
  id: Int
  edited: String
  climate: String
  surface_water: String
  name: String
  diameter: Int
  rotation_period: Int
  created: String
  terrain: String
  gravity: String
  orbital_period: Int
  population: Int
  residents: [Person]
  films: [Film]
}

type Species implements Node {
  edited: String
  classification: String
  name: String
  designation: String
  created: String
  eye_colors: String
  people: [Person]
  skin_colors: String
  language: String
  hair_colors: String
  homeworld: Planet
  average_lifespan: Int
  average_height: Int
  id: Int
}

type Transport implements Node {
  edited: String
  consumables: String
  name: String
  created: String
  cargo_capacity: Int
  passengers: Int
  max_atmosphering_speed: Int
  crew: String
  length: Int
  model: String
  cost_in_credits: Int
  manufacturer: String
  image: String
  id: Int
}

input AdvancedInput {
  enumField: NodeType!
  otherField: String!
}

directive @i18n(locale: String) on QUERY
"
`;
